src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/TimedFlipFlopDetector.java

```diff
@@ -0,0 +1,118 @@
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.decisionmaker.actions.ImpactVector.Dimension;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.decisionmaker.actions.ImpactVector.Impact;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.store.rca.cluster.NodeKey;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link FlipFlopDetector} whose recorded actions expire after a given period of time.
+ *
+ * <p>This class defines a flip flop as an {@link Impact#DECREASES_PRESSURE}s followed by an
+ * {@link Impact#INCREASES_PRESSURE}s to be a flip flops.
+ *
+ * <p>This class stores a {@link TimeExpiringSet} of {@link ImpactVector}s per {@link NodeKey}
+ * that are used to determine these flip flops.
+ */
+public class TimedFlipFlopDetector implements FlipFlopDetector {
+    private Map<NodeKey, TimeExpiringSet<ImpactVector>> flipFlopMap;
+    private long expiryDuration;
+    private TimeUnit expiryUnit;
+
+    public TimedFlipFlopDetector(long duration, TimeUnit unit) {
+        flipFlopMap = new HashMap<>();
+        this.expiryDuration = duration;
+        this.expiryUnit = unit;
+    }
+
+    /**
+     * Tests if (prev, curr) is a flip flopping sequence of impacts.
+     *
+     * <p>Only an increase following a decrease is considered a flip flop. Therefore, if
+     * prev decreases pressure and curr increases pressure, then (prev, curr) is a flip flop.
+     *
+     * @param prev The {@link Impact} that curr is compared against
+     * @param curr The {@link Impact} that you'd like to test and apply
+     * @return Whether or not (prev, curr) is a flip flopping sequence of impacts
+     */
+    protected boolean isFlipFlopImpact(Impact prev, Impact curr) {
+        return prev.equals(Impact.DECREASES_PRESSURE) && curr.equals(Impact.INCREASES_PRESSURE);
+    }
+
+    /**
+     * Returns true if the impact for any given Dimension in prev is a flip flop Impact when compared to
+     * the impact for a given dimension in prev
+     *
+     * <p>e.g. for prev = (HEAP: INCREASE, CPU: DECREASE), curr = (HEAP: DECREASE, CPU: INCREASE)
+     * (prev, curr) is a flip flop vector because a CPU: DECREASE followed by a CPU: INCREASE is a flip
+     * flop impact. Note that (HEAP: DECREASE) followed by (CPU: INCREASE) is not a flip flop
+     * because HEAP =/= CPU.
+     *
+     * @param prev The first {@link ImpactVector}. Its Impacts appear on the LHS of calls to
+     *             {@link this#isFlipFlopImpact(Impact, Impact)}
+     * @param curr The second {@link ImpactVector}. Its Impacts appear on the RHS of calls to
+     *            {@link this#isFlipFlopImpact(Impact, Impact)}.
+     * @return true if the impact for any given Dimension in curr is a flip flop Impact when compared to
+     *      the impact for a given dimension in prev
+     */
+    protected boolean isFlipFlopVector(ImpactVector prev, ImpactVector curr) {
```

Comment:
Do you think we will need the resource name for which the flip flop happened ? 
So, for `for prev = (HEAP: INCREASE, CPU: INCREASE), curr = (HEAP: DECREASE, CPU: INCREASE)`, we can return `HEAP` instead of Boolean value ?

What I want to understand is if we use(or might use in future) this information anywhere.

Reply:
Since it's trivial to add this information in the future by extracting the logic in this function, I think we can table this until the need arises